### ISSUE TITLE
[$60] - Fix the local deployment of the Arena

### ISSUE THREAD

**Issue by birdofpreyru** [appirio-tech/topcoder-app#957]
Clone https://github.com/appirio-tech/arena-web and try to deploy it locally according to `README.md`. You will see, that `grunt` command fails due to some broken dependencies. Your goal here is to fix the local deployment, so that it does deploy and works similarly to the production version at https://arena.topcoder.com. Make sure to modify `README.md` so that it is clear how to rapidly deploy the Arena locally.

**Comment by lijulat** [comment]
@birdofpreyru A quick question I have corrected the grunt errors and the application builds locally. When i try to browse the application it redirects me to the live login screen and after login it take me to live tc dashboard. Should we correct the code to fix this behaviour?

**Comment by birdofpreyru** [comment]
@lijulat Are you able to access the locally deployed Arena, if after the login you manually type its address into URL? If yes, then I'm okey with that (this behavior is not good, but the same happens when I'm accessing the production version of Arena without being logged in), but it should be mentioned in `README.md`.

Also, please verify that it is generally working properly. E.g. when you interact with that, asking to load problems, compile code, etc. it communicates with the proper remote API, and works similarly to the live version in prod.

**Comment by lijulat** [comment]
@birdofpreyru i am unable to verify because its configured to http://tc.cloud.topcoder.com/ which i think is a locally deployed version . Do you have details for the live api ?

**Comment by lijulat** [comment]
@birdofpreyru I would need the correct websocket url  currently it is http://tc.cloud.topcoder.com:5016.

I managed to get the other urls from the tc-site repo.

**Comment by birdofpreyru** [comment]
@lijulat Do you mean, you can't figure out yourself the correct URL for this socket?
@ajefts Do you know what does this correspond to, and what should be the correct URL?

**Comment by lijulat** [comment]
@birdofpreyru yes. Since all the urls in the config are not valid or refer to some locally deployed urls. Couple of the urls i was able to figure out the hosted ones from the tc-site config. But there is URL named websocket url which I am unable to figure out.

**Comment by birdofpreyru** [comment]
@lijulat Ok, let's wait for a reply from @ajefts, I guess he just knows what this URL should be.

**Comment by birdofpreyru** [comment]
@lijulat By the way, have you tried to load the production version of the Arena with Chrome development console open at the Network tab? It logs you all network requests from the page, and probably, you can find out from there what this address should be.

**Comment by lijulat** [comment]
@birdofpreyru I got the address of the production version . But It's giving me connection time out error.  May be it doesn't allow connections from local deployment ?

**Comment by birdofpreyru** [comment]
@lijulat I don't know :) If you tell me the address, I can also look into it a bit later.

**Comment by lijulat** [comment]
https://arenaws.topcoder.com

**Comment by birdofpreyru** [comment]
@lijulat Please, also push your current version of the code into your fork of the Arena repo, so that I can use it for testing.

**Comment by lijulat** [comment]
@birdofpreyru  https://github.com/lijulat/arena-web/

**Comment by lijulat** [comment]
@birdofpreyru  Did you manage to look at this? Anything pending from me? I guess we might need the correct config values for staging/dev

**Comment by birdofpreyru** [comment]
@lijulat I've looked into it briefly, saw it didn't work, and didn't do anything else yet.

**Comment by rainerlopez** [comment]
May II try this one? How is the progress here @birdofpreyru ? Tks

**Comment by birdofpreyru** [comment]
@rainerlopez Sure! Currently we have stuck with this one.

**Comment by rainerlopez** [comment]
1.
STATIC_FILE_HOST pointing to http://arena.cloud.topcoder.com:%PORT%
set STATIC_FILE_HOST=http://arena.cloud.topcoder.com:%PORT%
Just to situate myself, do we really need this dynamism intead of just <script src="js/newrelic.js"></ script> and <script src="js/bundle.js"></ script>?

2.
After everything installed it seems that the bundle.js file is missing inside /js/ folder.

**Comment by rainerlopez** [comment]
Copied production https://arena.topcoder.com/js/bundle.js to local /js/ folder and everything seems to be ok.
The problem now is that everything is pointing to production including the authentication.
How do you guys think we can procced from here?

Continuing with the investigations and tests while waiting for instructions.

**Comment by birdofpreyru** [comment]
@rainerlopez Yeap, like @lijulat said, it is exactly where we have stuck.

In principle, if there is a way to make it work against the production backend, I'll be fine with it for now. The core problem now is that after the locally deployed Arena redirects you to the prod login page, and you log in, you get redirected to the remote Arena page, not the local one.

**Comment by rainerlopez** [comment]
Nice ! Thanks @lijulat & @birdofpreyru 
I'll manage it and let you know ASAP.

**Comment by rainerlopez** [comment]
@birdofpreyru & @lijulat 
I'll need to fix this one also: https://github.com/appirio-tech/arena-web/issues/707

**Comment by birdofpreyru** [comment]
@rainerlopez I believe, @lijulat has already fixed it in his clone of the repo.

**Comment by rainerlopez** [comment]
Oh, I'll take a look ! Thanks @birdofpreyru

**Comment by birdofpreyru** [comment]
@lijulat @rainerlopez Have you seen this document inside the Arena repo?
https://github.com/appirio-tech/arena-web/wiki/Web-Arena-Setup-Guide-For-Ubuntu
I've just noticed it by chance, looking for some other staff, and had no time to look into details, but it looks like it might have the info necessary for the deployment.

**Comment by lijulat** [comment]
@birdofpreyru I will check and comeback to you. Initial glance looks promising.

**Comment by lijulat** [comment]
@birdofpreyru A quick question though. Who should be looking into this? Since this ticket was assigned to me I thought I was . Bit confused now.

**Comment by birdofpreyru** [comment]
Probably, you guys can collaborate? We'll split the prize according to efforts and contribution?

All I really care now, that we get the code fixed to work locally, and the deployment procedure well documented so that anybody with minimal experience can deploy it.

Once we have it, I believe, we'll run further challenges to enhance the Arena, so understanding how it works now, will give you some benefits in future ;)

**Comment by rainerlopez** [comment]
@birdofpreyru and @lijulat the document is promissing, yes.
But I think that r.now we need to fix the ng-table problem because it does not support bower anymore: https://github.com/esvit/ng-table Bower: NO LONGER SUPPORTED
And one of the instructions of the doc is: "bower install" and etc

It's giving me a hard time but I'm working on it.

@lijulat are you going to continue with this thread? If yes, no problem. We can try to chat and I'll try to help you with some intel I've discovered till now. If not, I'll try to solve everything.

Thanks guys

**Comment by lijulat** [comment]
@birdofpreyru Looks like from the wiki we need an Arena VM for the local deployment to work. That is where the webscoket connections point to. Excerpts from the wiki below for quick reference

"Upon registration as a submitter or reviewer you will need to request an Arena VM in order to make Web Arena functioning properly. To request your image, please post in contest forum.

Before requesting your VM, you need to ensure that you have an SSH key created and in your member profile. Instructions to do so are here:

http://www.topcoder.com/wiki/display/projects/Generate+SSH+Key,

And instructions to connect afterwards are here:

http://www.topcoder.com/wiki/display/projects/Connect+Using+SSH+Key.

Once you get your VM IP, you need to add the following line in your hosts file as shown in section 2.13

<<vm-ip>>   tc.cloud.topcoder.com"

**Comment by lijulat** [comment]
@rainerlopez NG table used by arena web production is v0.4.1 and I think bower supports version till v0.8.3. So that shouldn't be a problem

**Comment by rainerlopez** [comment]
@lijulat you are right. But I think we can/need to fix it to be able to follow the library updates and etc.
By now, if we want to stick to any version we'll need to change bower.json line 28:
"angular-table": "https://github.com/esvit/ng-table.git",
that is pointing to their git and is breaking the "grunt" browserify task.

@birdofpreyru can give us some instructions on what to do by now with this issue.

About the docs @lijulat is correct, @birdofpreyru. Can we get the VM ?

**Comment by lijulat** [comment]
@rainerlopez I think you are referring to wrong repo. I had already fixed this in my repo couple of weeks back. There are no npm or bower install errors in my repo. This has been verified by @birdofpreyru  as well.

**Comment by rainerlopez** [comment]
@lijulat that's great, I'll clone yours and keep investigating. Any improvement I'll let you know.
@birdofpreyru I think my collaboration here is a community contribution and Liju diserves the prize as soon as we finish all the requirements.
Count on me for now on.

**Comment by birdofpreyru** [comment]
As I remember it, @lijulat indeed has fixed it, so that building process works without errors in his code. The first problem it has is that it redirects to the prod. authentication App, and from there it redirects to the prod. Arena (or just to prod. TC dashboard - I don't remember exactly now). Maybe there some further problems with the code deployment, but we can't see them yet.

Ok, I've noted about VM. I don't know yet how and where to get it, but I'll figure it out.

**Comment by rainerlopez** [comment]
@birdofpreyru I think that with the VM and following the doc instructions the local deploy will be smooth.
The url's point for the VM and so the socket.
@lijulat is ahead with this and will probably solve everything as soon as he get the VM.

**Comment by birdofpreyru** [comment]
Just to be sure, are you positive, that there is no way to connect it to the production backend with the data we get? It would be easier to deploy, and I believe there are numerous enhancements which can be done with the frontend Arena App, without changing the backend.

**Comment by rainerlopez** [comment]
@birdofpreyru No!
Yes we can point directly to the production. (That will change the way it is set r.now. Maybe we don't even need the VM)
If this is you final expectations @lijulat or myself can work on it.
Of course, it will be a bit more work.

What do you think @lijulat ?

**Comment by birdofpreyru** [comment]
Yeah, if we get the locally deployed frontend to work fine against the production backend, it will be fine for now. I believe, re-configuring it to the local backend in VM won't take much efforts, and also we will be able to start working on Arena App without waiting till we figure out about VM.

**Comment by rainerlopez** [comment]
Great, I'll start investigating here.
@lijulat as soon as you read this thread please call me and let's sync to not do the same thing we both.

**Comment by lijulat** [comment]
@rainerlopez You can continue the investigation. I am currently looking into something else.

**Comment by rainerlopez** [comment]
@birdofpreyru is there someone who can confirm if the prod socket (webSocketURL: 'https://arenaws.topcoder.com',) accept remote connections ?

**Comment by birdofpreyru** [comment]
I guess, it is supposed to, as the prod Arena does work somehow.
I think, nobody will confirm the details rapidly. It seems that Arena was deployed last time over an year ago, so it is hard to get the details now.

**Comment by rainerlopez** [comment]
Don't worry, I'll test it till exaustion ;)

**Comment by rainerlopez** [comment]
Hi @birdofpreyru & @lijulat 
Well, I think we will need some help here:
We have some work to be done to point directly to production. Some tasks that I've listed in my investigation:

- Facebook App: We need to list the local domain in the app settings
- Websocket: We will need someone (at the dark side of the servers) who can help us telling what need to be done to manage the remote connection.

In the other hand, if we get the VM we can still try to make it work locally.

I'll keep investigating today and update you guys.

**Comment by birdofpreyru** [comment]
FYI: A copilot more knowledgeble about TC backend is taking care about local deployment of different parts of TC site. I am not closely following on his plan&progress, but I see, he has just run a F2F challenge to deal with the login problem: https://www.topcoder.com/challenge-details/30056747/?type=develop&nocache=true You guys might want to participate, as he also have a better knowledge about configurations etc.
